### PR TITLE
Reject generic self types.

### PR DESCRIPTION
**Pull request by adetaylor**
The RFC for arbitrary self types v2 declares that we should reject "generic" self types. This commit does so.

The definition of "generic" was unclear in the RFC, but has been explored in
https://github.com/rust-lang/rust/issues/129147
and the conclusion is that "generic" means any `self` type which is a type parameter defined on the method itself, or references to such a type.

This approach was chosen because other definitions of "generic" don't work. Specifically,
* we can't filter out generic type _arguments_, because that would filter out Rc<Self> and all the other types of smart pointer we want to support;
* we can't filter out all type params, because Self itself is a type param, and because existing Rust code depends on other type params declared on the type (as opposed to the method).

This PR decides to make a new error code for this case, instead of reusing the existing E0307 error. This makes the code a bit more complex, but it seems we have an opportunity to provide specific diagnostics for this case so we should do so.

This PR filters out generic self types whether or not the 'arbitrary self types' feature is enabled. However, it's believed that it can't have any effect on code which uses stable Rust, since there are no stable traits which can be used to indicate a valid generic receiver type, and thus it would have been impossible to write code which could trigger this new error case. It is however possible that this could break existing code which uses either of the unstable `arbitrary_self_types` or `receiver_trait` features. This breakage is intentional; as we move arbitrary self types towards stabilization we don't want to continue to support generic such types.

This PR adds lots of extra tests to arbitrary-self-from-method-substs. Most of these are ways to trigger a "type mismatch" error which https://github.com/rust-lang/rust/blob/9b82580c7347f800c2550e6719e4218a60a80b28/compiler/rustc_hir_typeck/src/method/confirm.rs#L519 hopes can be minimized by filtering out generics in this way. We remove a FIXME from confirm.rs suggesting that we make this change. It's still possible to cause type mismatch errors, and a subsequent PR may be able to improve diagnostics in this area, but it's harder to cause these errors without contrived uses of the turbofish.

This is a part of the arbitrary self types v2 project, https://github.com/rust-lang/rfcs/pull/3519
https://github.com/rust-lang/rust/issues/44874

r? @wesleywiser

### CODE REVIEW

**Comment by adetaylor** [comment]
@wesleywiser Over to you for opinions here. See [this comment](https://github.com/rust-lang/rust/issues/129147#issuecomment-2337637087) for why I think this is the least bad approach for filtering out generics - to be explicit, we know it's still possible to have generic self types, but this probably reduces the risk of them being _common_ and also resolves a `FIXME`.

I'd appreciate your help and adjudication about whether this is the right overall approach. We should probably get to the bottom of that before you review the PR in detail, but it's up to you!

@rustbot label -S-waiting-on-author +S-waiting-on-review

File: compiler/rustc_hir_analysis/src/check/wfcheck.rs [hunk 1]
```diff
@@ -1750,13 +1765,53 @@ fn check_method_receiver<'tcx>(
...
+    method_generics: Option<&hir::Generics<'_>>,
+) -> Result<(), ReceiverValidityError> {
+    if let ty::Param(param) = ty.kind() {
+        if let Some(generics) = method_generics {
+            if generics
+                .params
+                .iter()
+                .filter(|g| matches!(g.kind, GenericParamKind::Type { .. }))
+                .map(|g| g.name.ident().name)
+                .contains(&param.name)
```
Comment (wesleywiser) [inline compiler/rustc_hir_analysis/src/check/wfcheck.rs:1806]: @compiler-errors is checking if the `ty::Param` name matches the name of a generic parameter on the method the right way to do this check?

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/wfcheck.rs:1806]: Definitely not lol, that is not hygienic and also unnecessarily iterating through all the generics.

Comment (compiler-errors) [reply compiler/rustc_hir_analysis/src/check/wfcheck.rs:1806]: Just compare the index of the param to the `parent_count`: https://doc.rust-lang.org/stable/nightly-rustc/rustc_middle/ty/generics/struct.Generics.html

**Comment by wesleywiser** [comment]
> See https://github.com/rust-lang/rust/issues/129147#issuecomment-2337637087 for why I think this is the least bad approach for filtering out generics - to be explicit, we know it's still possible to have generic self types, but this probably reduces the risk of them being common and also resolves a FIXME.

@adetaylor Thanks for that link! After reading that thread and a few others I stumbled across along the way, I think the approach makes sense and your implementation looks fine to me (I had one question I was unsure of and pinged compiler-errors for their opinion). The additions you made to the test suite were also really helpful in my initial understanding and I'm happy to see more test coverage there!

File: compiler/rustc_hir_analysis/src/check/wfcheck.rs (hunk 1 above)
Comment (adetaylor) [reply compiler/rustc_hir_analysis/src/check/wfcheck.rs]: I had a crack at this in c2cbf1d8848d9bacb4a3035646405a57b5f90e73.

**Review by wesleywiser — APPROVED** [review]

**Comment by wesleywiser** [comment]
This looks great to me, just needs a rebase. 

@bors delegate+

@adetaylor You can `@bors r=wesleywiser` after the merge conflict is fixed 🙂

**Comment by wesleywiser** [comment]
Sigh, I thought it would ignore the command since it didn't start the line. 

@bors r-

**Pull request merged**